keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,350 @@
+	// information to pass to the CSI driver to complete the CSI NodeStageVolume
+	// and NodeStageVolume and NodeUnstageVolume calls.
+	// This field is optional, and  may be empty if no secret is required. If the
+	// secret object contains more than one secret, all secret references are passed.
+	// 
+	// +optional
+	NodeStageSecretRef *LocalObjectReference
+
+	// NodePublishSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// NodePublishVolume and NodeUnpublishVolume calls.
+	// This field is optional, and  may be empty if no secret is required. If the
+	// secret object contains more than one secret, all secret references are passed.
+	// 
+	// +optional
+	NodePublishSecretRef *LocalObjectReference
+}
+```
+
+### Secret references
+All secret references declared in a CSI inline volume can only refer to pre-defined secrets within the same namespace where the corresponding pod is running.  Secret references are stored in `LocalObjectReference` instances which require a pointer to the secret and a namespace value.
+
+### Specifying allowed inline drivers with `PodSecurityPolicy`
+To control which driver is allowed to be used within a pod spec, this design will update the `PodSecurityPolicy` to introduce `AllowedCSIDrivers` as shown below:
+
+
+```go
+  type PodSecurityPolicySpec struct {
+	// <snip>
+
+	// AllowedCSIDrivers is a whitelist of allowed CSI drivers.  Empty or nil indicates that all
+	// CSI drivers may be used.  This parameter is effective only when the usage of the CSI plugin
+	// is allowed in the "Volumes" field.
+	// +optional
+	AllowedCSIDrivers []AllowedCSIDriver
+  }
+
+  // AllowedCSIDriver represents a single CSI driver that is allowed to be used.
+  type AllowedCSIDriver struct {
+	// Name of the CSI driver
+	Name string
+  }  
+```
+
+Value `PodSecurityPolicy.AllowedCSIDrivers` must be explicitly set with the names of CSI drivers that are allowed to be embedded within a pod spec.  An empty value means no CSI drivers are allowed to be specified inline inside a pod spec.
+
+### Persistent vs ephemeral lifecycle setting
+A CSI driver must indicate that it supports either persistent or ephemeral inline mode.  This setting will be stored in a [`CSIDriver` configuration CRD](https://github.com/kubernetes/community/pull/2514) and must be retrieved from the API server at runtime.
+
+#### Persistent inline mode
+For CSI drivers that support PV/PVC, the `CSIDriver.lifecycle` value should be set to `persistent` (default).  That value indicates the following:
+
+* The driver uses the normal storage operation flow of *provision*, *attach*,*mount*, *unmount*, *detach*, *delete*.
+* When used in an inline context: 
+  * The `volumeHandle` value must be generated ahead of time and manually entered in the spec 
+  * The driver is only responsible for *attach*, *mount*, *unmount*, *detach*
+  * Provision and deletion are handled outside of the driver
```

Comment:
Also mention all the other PVC/PV features that won't work:
- topology
- raw block
- mount options
- resizing

Reply:
Added to non-goal